mypy/constraints.py

```diff
@@ -228,18 +228,13 @@ def infer_constraints_for_callable(
                 if param_spec and callee.arg_kinds[i] in (ARG_STAR, ARG_STAR2):
                     # If actual arguments are mapped to ParamSpec type, we can't infer individual
                     # constraints, instead store them and infer single constraint at the end.
-                    # It is impossible to map actual kind to formal kind, so use some heuristic.
-                    # This inference is used as a fallback, so relying on heuristic should be OK.
                     if not incomplete_star_mapping:
                         param_spec_arg_types.append(
                             mapper.expand_actual_type(
                                 actual_arg_type, arg_kinds[actual], None, arg_kinds[actual]
                             )
                         )
-                        actual_kind = arg_kinds[actual]
-                        param_spec_arg_kinds.append(
-                            ARG_POS if actual_kind not in (ARG_STAR, ARG_STAR2) else actual_kind
-                        )
+                        param_spec_arg_kinds.append(arg_kinds[actual])
```

Comment:
@ilevkivskyi I made this change and the only fallout from the test cases seems to be lambda inference (and error messages. Oh well...). I think that's probably a solvable problem, so I'm curious whether I'm missing some case.

Reply:
I am not only worried about the test cases, the bigger problem is that this change doesn't make much sense to me (read again the comment you deleted). What you are trying to achieve should be done by handling unions here https://github.com/python/mypy/blob/master/mypy/checkexpr.py#L2281-L2291.

Reply:
Yeah that sounds like a better spot to modify.

Just to confirm vocabulary, formal kind is referring to the kind within the (unknown) parameters? If so then I figured that since this is a supertype things will work out compatibility-wise.

I mean anyways it's easier to modify where you linked, so I'll do that and revert this change.